qualtran/bloqs/factoring/mod_arithmetic.py

```diff
@@ -0,0 +1,224 @@
+from qualtran.resource_counting import BloqCountT, SympySymbolAllocator
+
+
+@frozen
+class ModAdd(Bloq):
+    r"""An n-bit modular addition gate.
+
+    Implements $U|x\rangle|y\rangle|p\rangle \rightarrow |x\rangle|y + x mod p\rangle|p\rangle$ using $4n T$ gates.
+
+    Args:
+        bitsize: Number of bits used to represent each integer.
+
+    Registers:
+        x: A bitsize-sized input register (register x above).
+        y: A bitsize-sized input/output register (register y above).
+        p: A bitsize-sized input register (register p above).
+
+    References:
+        [How to compute a 256-bit elliptic curve private key with only 50 million Toffoli gates](https://arxiv.org/abs/2306.08585)
+    """
+
+    bitsize: Union[int, sympy.Expr]
+
+    @cached_property
+    def signature(self) -> 'Signature':
+        return Signature(
+            [
+                Register('x', bitsize=self.bitsize),
+                Register('y', bitsize=self.bitsize),
+                Register('p', bitsize=self.bitsize),
+            ]
+        )
+
+    def bloq_counts(self, ssa: Optional['SympySymbolAllocator'] = None) -> Set['BloqCountT']:
+        return {(4 * self.bitsize, Toffoli())}
+
+    def build_composite_bloq(
+        self, bb: 'BloqBuilder', x: SoquetT, y: SoquetT, p: SoquetT
+    ) -> Dict[str, 'SoquetT']:
+        return NotImplemented
+
+    def short_name(self) -> str:
+        return f'y = y + x mod p'
+
+
+@frozen
+class ModSub(Bloq):
+    r"""An n-bit modular subtraction gate.
+
+    Implements $U|x\rangle|y\rangle|p\rangle \rightarrow |x\rangle|y - x mod p\rangle|p\rangle$ using $6n T$ gates.
+
+    Args:
+        bitsize: Number of bits used to represent each integer.
+
+    Registers:
+        x: A bitsize-sized input register (register x above).
+        y: A bitsize-sized input/output register (register y above).
+        p: A bitsize-sized input register (register p above).
+
+    References:
+        [How to compute a 256-bit elliptic curve private key with only 50 million Toffoli gates](https://arxiv.org/abs/2306.08585)
+    """
+
+    bitsize: Union[int, sympy.Expr]
+
+    @cached_property
+    def signature(self) -> 'Signature':
+        return Signature(
+            [
+                Register('x', bitsize=self.bitsize),
+                Register('y', bitsize=self.bitsize),
+                Register('p', bitsize=self.bitsize),
+            ]
+        )
+
+    def bloq_counts(self, ssa: Optional['SympySymbolAllocator'] = None) -> Set['BloqCountT']:
+        return {(6 * self.bitsize, Toffoli())}
+
+    def build_composite_bloq(
+        self, bb: 'BloqBuilder', x: SoquetT, y: SoquetT, p: SoquetT
+    ) -> Dict[str, 'SoquetT']:
+        return NotImplemented
+
+    def short_name(self) -> str:
+        return f'y = y - x mod p'
+
+
+@frozen
+class ModNeg(Bloq):
+    r"""An n-bit modular negation gate.
+
+    Implements $U|x\rangle|p\rangle \rightarrow |-x mod p\rangle|p\rangle$ using $2n T$ gates.
```

Comment:
Is it T gates or Toffolis? The bloq_counts has 2 n Toffoli gates, a Toffoli can be decomposed into 4 T gates.

Reply:
It looks like the paper quotes Toffolis so the counts are correct, just change T -> Toffoli in the docstrings.